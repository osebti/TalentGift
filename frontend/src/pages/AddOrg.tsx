//#region import
import config from "../../../config";

// components
import Button from "@/components/Button";
import TextBox from "@/components/TextBox";
import { setGlobalState, useGlobalState } from "@/state/global";
import { useRef, useState } from "react";
import { Link, useNavigate } from "react-router-dom";
//#endregion

export default function AddOrg() {
  const [user] = useGlobalState("user");
  const [org] = useGlobalState("org");
  const orgNameRef = useRef<HTMLInputElement>(null);
  const [validInput, setValid] = useState(true);
  const navigate = useNavigate();

  // handle sending a request to the server to create an organization
  const createOrgOnClick = () => {
    if (orgNameRef.current != null) {
      const orgName = orgNameRef.current.value.trim();

      // input checking
      if (orgName === "" || orgName.length > 255) {
        setValid(false);
      }
      
      // otherwise send the request
      else {
        fetch(`http://${config.hostname}/organization/create_organization`, {
        method: "POST",
        headers: {
          "Accept": "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          mid: user.uid,
          orgname: orgName,
        }),
      })
        .then((res) => res.json())
        .then((data) => {
          // navigate to success page if status is 0
          if (data.status === 0) {
            navigate("/success", {
              state: {
                message: "Successfully created organization!",
                returnAddress: "/dashboard",
                returnName: "Dashboard",
              },
            });
            setGlobalState("user", { ...user, manager: true }); // denote that user is now a manager
            setGlobalState("org", { ...org, oid: data.id });
            // nativate to error page if status is 1
          } else if (data.status === 1) {
            navigate("/error", {
              state: {
                message: "Failed to create an organization ...",
                returnAddress: "/dashboard",
                returnName: "Dashboard",
              },
            });
          }
        })
        .catch((err) => {
          // navigate to error page if exception occurred
          navigate("/error", {
            state: {
              message: err,
              returnAddress: "/dashboard",
              returnName: "Dashboard",
            },
          });
        });
      }
    }
  };

  // Add Organization page
  return (
    <div className="w-full h-full flex justify-center items-center px-4">
      <div className="flex flex-col w-2/3 justify-center items-center gap-2">
        {/** Title and subtitle of the page */}
        <div
          id="content-header"
          className="flex flex-col justify-center items-center my-6 md:my-8 lg:my-10"
        >
          <div
            id="content-title"
            className="font-semibold text-2xl md:text-3xl lg:text-4xl text-black"
          >
            Set up your organization
          </div>
          <div
            id="content-subtitle"
            className="text-md md:text-lg lg:text-xl text-secondary"
          >
            Tell us about your organization
          </div>
        </div>
        {/** Organization creation form */}
        <div
          id="create-org-form"
          className="flex flex-col w-full justify-center items-start gap-6"
        >
          {/** Organization name input field and invalid input message */}
          <TextBox
            id="org-name"
            label=""
            placeholder="Organization Name"
            type="text"
            textRef={orgNameRef}
          />
          { validInput 
            ? null
            : <p className="text-error">
                Invalid input. Enter a non-empty name at most 255 characters in length.
              </p>
          }
          {/** Create organization button */}
          <Button
            id="create-org"
            label="Next"
            style="btn-secondary w-full rounded-xl"
            onClick={createOrgOnClick}
          />
        </div>
        {/** Back to dashboard link */}
        <Link className="mx-4 md:mx-8 lg:mx-0" to="/dashboard">
          <div className="flex justify-center items-center">
            <div className="text-secondary hover:text-black transition ease-in-out text-center">
              I change my mind, back to Dashboard
            </div>
          </div>
        </Link>
      </div>
    </div>
  );
}
