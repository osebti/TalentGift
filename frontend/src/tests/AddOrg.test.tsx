//#region imports
import { cleanup, fireEvent, render, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import fetchMock from 'jest-fetch-mock';

import AddOrg from "@/pages/AddOrg";
import { BrowserRouter } from "react-router-dom";
//#endregion

//#region test setup
fetchMock.enableMocks();
console.error = jest.fn();
console.warn = jest.fn();
console.log = jest.fn();

beforeEach(() => {
    fetchMock.resetMocks();
});

afterEach(cleanup);
//#endregion

//#region functions and objects used for testing
function renderAddOrgPage() {
  return (
    render(
      <BrowserRouter>
        <AddOrg/>
      </BrowserRouter>
    )
  );
}
//#endregion

//#region tests
describe("Rendering the AddOrg Page", () => {

  it("renders correctly", () => {
    const page = renderAddOrgPage();
    expect(page).toMatchSnapshot();
  });

  it("renders organization name input field", () => {
    const { queryByTestId } = renderAddOrgPage();
    expect(queryByTestId("org-name textbox")).not.toBeNull();
  });

  it("renders create organization button", () => {
    const { queryByTestId } = renderAddOrgPage();
    expect(queryByTestId("create-org-button")).not.toBeNull();
  });

});

describe("Creating an organization", () => {

  it("handles successful organization creation", async () => {
    const { queryByTestId } = renderAddOrgPage();
    
    // mock successful organization creation request
    fetchMock.mockResponseOnce(
      JSON.stringify({ status: 0 }), 
      { status: 200, headers: { 'Content-Type': 'application/json' } }
    );

    // enter an organization name and submit
    const nameInput = queryByTestId("org-name textbox");
    const createOrgBtn = queryByTestId("create-org-button");

    if (nameInput != null && createOrgBtn != null) {
      fireEvent.change(nameInput, { target: { value: "Test Organization Name" } });
      fireEvent.click(createOrgBtn);
    }

    // check if fetch request was sent and page was redirected
    await waitFor(() => {
      expect(fetchMock).toHaveBeenCalledTimes(1);
      expect(global.window.location.href).toContain("/success");
    });
  });

  it("does not create organization if no organization name is given", () => {
    const { queryByTestId, queryByText } = renderAddOrgPage();

    // check that the invalid input message is not already showing
    expect(queryByText("Invalid input. Enter a non-empty name at most 255 characters in length.")).toBeNull();

    // press submit without entering an organization name
    const createOrgBtn = queryByTestId("create-org-button");

    if (createOrgBtn != null) {
      fireEvent.click(createOrgBtn);
    }

    // check if the invalid input message is shown
    expect(queryByText("Invalid input. Enter a non-empty name at most 255 characters in length.")).not.toBeNull();
  });

});
//#endregion