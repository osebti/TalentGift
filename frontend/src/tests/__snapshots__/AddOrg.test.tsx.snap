// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering the AddOrg Page renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="w-full h-full flex justify-center items-center px-4"
      >
        <div
          class="flex flex-col w-2/3 justify-center items-center gap-2"
        >
          <div
            class="flex flex-col justify-center items-center my-6 md:my-8 lg:my-10"
            id="content-header"
          >
            <div
              class="font-semibold text-2xl md:text-3xl lg:text-4xl text-black"
              id="content-title"
            >
              Set up your organization
            </div>
            <div
              class="text-md md:text-lg lg:text-xl text-secondary"
              id="content-subtitle"
            >
              Tell us about your organization
            </div>
          </div>
          <div
            class="flex flex-col w-full justify-center items-start gap-6"
            id="create-org-form"
          >
            <div
              class="flex-col justify-start items-start w-full"
              id="org-name-container"
            >
              <div
                class="text-start text-black text-xl font-normal pb-2"
                data-testid="org-name header"
                id="org-name-header"
              />
              <input
                class="text-black input outline outline-gray-300 text-xl font-normal bg-transparent w-full px-3 py-2.5 rounded-md justify-start items-center inline-flex focus:outline-darkPurple"
                data-testid="org-name textbox"
                id="org-name-textbox"
                placeholder="Organization Name"
                type="text"
              />
            </div>
            <button
              class="btn btn-md capitalize font-normal text-white text-sm md:text-base btn-secondary w-full rounded-xl"
              data-testid="create-org-button"
              id="create-org-button"
            >
              Next
            </button>
          </div>
          <a
            class="mx-4 md:mx-8 lg:mx-0"
            href="/dashboard"
          >
            <div
              class="flex justify-center items-center"
            >
              <div
                class="text-secondary hover:text-black transition ease-in-out text-center"
              >
                I change my mind, back to Dashboard
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="w-full h-full flex justify-center items-center px-4"
    >
      <div
        class="flex flex-col w-2/3 justify-center items-center gap-2"
      >
        <div
          class="flex flex-col justify-center items-center my-6 md:my-8 lg:my-10"
          id="content-header"
        >
          <div
            class="font-semibold text-2xl md:text-3xl lg:text-4xl text-black"
            id="content-title"
          >
            Set up your organization
          </div>
          <div
            class="text-md md:text-lg lg:text-xl text-secondary"
            id="content-subtitle"
          >
            Tell us about your organization
          </div>
        </div>
        <div
          class="flex flex-col w-full justify-center items-start gap-6"
          id="create-org-form"
        >
          <div
            class="flex-col justify-start items-start w-full"
            id="org-name-container"
          >
            <div
              class="text-start text-black text-xl font-normal pb-2"
              data-testid="org-name header"
              id="org-name-header"
            />
            <input
              class="text-black input outline outline-gray-300 text-xl font-normal bg-transparent w-full px-3 py-2.5 rounded-md justify-start items-center inline-flex focus:outline-darkPurple"
              data-testid="org-name textbox"
              id="org-name-textbox"
              placeholder="Organization Name"
              type="text"
            />
          </div>
          <button
            class="btn btn-md capitalize font-normal text-white text-sm md:text-base btn-secondary w-full rounded-xl"
            data-testid="create-org-button"
            id="create-org-button"
          >
            Next
          </button>
        </div>
        <a
          class="mx-4 md:mx-8 lg:mx-0"
          href="/dashboard"
        >
          <div
            class="flex justify-center items-center"
          >
            <div
              class="text-secondary hover:text-black transition ease-in-out text-center"
            >
              I change my mind, back to Dashboard
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
